[AMDGPU] Change SGPR layout to striped caller/callee saved

shiltian opened the pull request:
This PR updates the SGPR layout to a striped caller/callee-saved design, similar
to the VGPR layout.

To ensure that s30-s31 (return address), s32 (stack pointer), s33 (frame
pointer), and s34 (base pointer) remain callee-saved, the striped layout starts
from s40, with a stripe width of 8. The last stripe is 10 wide instead of 8 to
avoid ending with a 2-wide stripe.

Fixes #113782.

shiltian commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/127353?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#127353** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/127353?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/127353?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#127352** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/127352?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

shiltian commented:
This has passed internal PSDB (except the one test case that I has not updated yet).

Flakebi reviewed (commented):
Sounds good to me

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallingConv.td:
```diff
@@ -91,7 +91,11 @@ def CSR_AMDGPU_AGPRs : CalleeSavedRegs<
+       (sequence "SGPR%u", 46, 53),
+       (sequence "SGPR%u", 62, 69),
+       (sequence "SGPR%u", 78, 85),
+       (sequence "SGPR%u", 94, 105))
```
This isn't quite 8, maybe adjust the description to be more specific? Also should comment the scheme here. Also, I can never remember what level of SGPRs corresponds to what occupancy

shiltian replied:
Yeah, I was thinking that as well. Maybe I can start with s32 and leave s30-s31 as caller saved?

shiltian replied:
However, s30-s31 are for return address. [A patch](https://reviews.llvm.org/D114652) made them callee saved years ago. I can start from s40 though. However, s32, s33, as well as s34 are used for stack pointer, frame pointer, and base pointer, respectively. Is it a good idea to make them caller saved?

cdevadas replied:
No. Do not change the behavior of these special registers s30 - s34. Yes, start making the CSR partition from s40.

arsenm replied:
We should be moving the special inputs to all be in callee saved sgprs. It's most critical that the return address be a callee save

cdevadas commented:
This patch would improve the codegen with less number of SGPR spills for heavy workloads involving device calls. May be run the perf PSDB as well? That would give us some initial numbers. You can CP this PR to the staging compiler and then launch the perf PSDB.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUCallingConv.td:
```diff
@@ -91,7 +91,11 @@ def CSR_AMDGPU_AGPRs : CalleeSavedRegs<
+       (sequence "SGPR%u", 94, 105))
 >;
 
 def CSR_AMDGPU_SI_Gfx_SGPRs : CalleeSavedRegs<
```
@Flakebi should we make some similar change here for amdgpu_gfx?

Flakebi replied:
I think both is fine (changing it or leaving it for now). amdgpu_gfx already has caller-saves that are not used for arguments, so it’s not hit by this bug.

The important part is that amdgpu_gfx wants the SGPR arguments to be in callee-save registers. I assume compute would likely benefit from having SGPR args in callee-saves as well, as they usually contain constant data, but it’s not there yet.
Once the C calling convention does that, we can probably ditch amdgpu_gfx and switch to the C calling conv for graphics.

jayfoad reviewed (commented):
Striping SGPRs serves no purpose on GFX10+ where all waves get the full allocation of SGPRs. But hopefully it doesn't do any harm either.

shiltian commented:
> Striping SGPRs serves no purpose on GFX10+ where all waves get the full allocation of SGPRs. But hopefully it doesn't do any harm either.

This doesn't serve as a performance improvement anyway. I'll request a performance cycle.

jayfoad commented:
> > Striping SGPRs serves no purpose on GFX10+ where all waves get the full allocation of SGPRs. But hopefully it doesn't do any harm either.
> 
> This doesn't serve as a performance improvement anyway. I'll request a performance cycle.

The only reason for doing striping is to get roughly the same ratio of callee-saves / non-callee-saves at different occupancies. Why would you want to keep that ratio constant, if not for performance?

Anyway that reason does not apply on GFX10+.

shiltian committed:
[AMDGPU] Change SGPR layout to striped caller/callee saved

This PR updates the SGPR layout to a striped caller/callee-saved design, similar
to the VGPR layout. The stripe width is set to 8.

Fixes #113782.

shiltian committed:
Start the partition from s40

shiltian commented:
> This patch would improve the codegen with less number of SGPR spills for heavy workloads involving device calls. May be run the perf PSDB as well? That would give us some initial numbers. You can CP this PR to the staging compiler and then launch the perf PSDB.

Will do.

arsenm commented:
Is the test from #113782 buried somewhere in this giant test diff?

shiltian commented:
> Is the test from #113782 buried somewhere in this giant test diff?

Yes. The check line of the test has been updated because it no longer crashes but another error was emitted. I'll fix the new issue in a follow up.

shiltian commented:
A full testing cycle has been requested. Will comment here afterwards.

shiltian commented:
Just got the results from a full cycle. There are no correctness issues and no performance regressions. I'm not sure if there's any performance improvement, though. That being said, this PR should be in good shape to go.

shiltian committed:
rebase and fix conflicts

rovka commented:
Hi @shiltian, I think this change introduced some failures on [this buildbot](https://lab.llvm.org/buildbot/#/builders/16/builds/15106), which went unnoticed because it was [already red](https://lab.llvm.org/buildbot/#/builders/16/builds/15105). Could you please have a look? :)

mikaelholmen commented:
So, with this patch e.g.
```
llc -verify-machineinstrs -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 < test/CodeGen/AMDGPU/materialize-frame-index-sgpr.gfx10.ll
```
fails with
```
# After Greedy Register Allocator
********** INTERVALS **********
SGPR59_LO16 [96r,96d:0) 0@96r
SGPR59_HI16 [96r,96d:0) 0@96r
%8 [16r,32r:0) 0@16r  weight:INF
%15 [80r,80d:0) 0@80r  weight:INF
RegMasks:
********** MACHINEINSTRS **********
# Machine code for function scalar_mov_materializes_frame_index_unavailable_scc: NoPHIs, TracksLiveness, TiedOpsRewritten, TracksDebugUserValues
Frame Objects:
  fi#0: size=16384, align=64, at location [SP]
  fi#1: size=4, align=4, at location [SP]

0B	bb.0 (%ir-block.0):
16B	  %8:vgpr_32 = V_MOV_B32_e32 %stack.0.alloca0, implicit $exec
32B	  INLINEASM &"; use alloca0 $0" [sideeffect] [attdialect], $0:[reguse:VGPR_32], %8:vgpr_32
80B	  dead renamable $sgpr4_sgpr5 = S_AND_B64 0, $exec, implicit-def $scc
96B	  $sgpr59 = S_MOV_B32 %stack.1.alloca1
112B	  INLINEASM &"; use $0, $1" [sideeffect] [attdialect], $0:[reguse], $sgpr59, $1:[reguse], killed $scc
128B	  SI_RETURN

# End machine code for function scalar_mov_materializes_frame_index_unavailable_scc.

*** Bad machine code: No live segment at use ***
- function:    scalar_mov_materializes_frame_index_unavailable_scc
- basic block: %bb.0  (0x55ed7ac21b50) [0B;144B)
- instruction: 112B	INLINEASM &"; use $0, $1" [sideeffect] [attdialect], $0:[reguse], $sgpr59, $1:[reguse], killed $scc
- operand 3:   $sgpr59
- liverange:   [96r,96d:0) 0@96r
- regunit:     SGPR59_LO16
- at:          112B

*** Bad machine code: No live segment at use ***
- function:    scalar_mov_materializes_frame_index_unavailable_scc
- basic block: %bb.0  (0x55ed7ac21b50) [0B;144B)
- instruction: 112B	INLINEASM &"; use $0, $1" [sideeffect] [attdialect], $0:[reguse], $sgpr59, $1:[reguse], killed $scc
- operand 3:   $sgpr59
- liverange:   [96r,96d:0) 0@96r
- regunit:     SGPR59_HI16
- at:          112B
LLVM ERROR: Found 2 machine code errors.
```
(also seen if you compile with EXPENSIVE_CHECKS and run lit tests, as the failed build bot shows)

shiltian commented:
@mikaelholmen @rovka Thanks for the information. I'll take a look right away.

shiltian commented:
Right before this PR, `CodeGen/AMDGPU/shufflevector-physreg-copy.ll` already fails with expensive checks. I'll fix the two new failures in https://github.com/llvm/llvm-project/pull/130644.
